File: src-tauri/src/repo/service.rs
//! Repo-state service: the single source of truth for "when does state refresh".
//!
//! One coordinator task per watched repo turns classified invalidations (from
//! the file watcher or from Grove's own write commands) into recomputed state
//! slices, compares them against the last emitted values, and pushes typed
//! `repo-event` payloads to the frontend. Every payload carries a generation
//! number so the frontend can drop anything stale. Nothing else in the app
//! re-fetches on its own; every refresh trigger funnels through `poke`.

use super::read;
use super::{CommitNode, WorkingStatus, Worktree};
use serde::Serialize;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;
use std::time::{Duration, Instant};
use tauri::Emitter;
use tokio::sync::mpsc;

/// Invalidation bits: which parts of repo state a change may have touched.
pub const INV_REFS: u8 = 1 << 0; // .git/refs, HEAD, packed-refs → graph, branches, unpushed
pub const INV_INDEX: u8 = 1 << 1; // .git/index → staged status (agents staging files)
pub const INV_WORKDIR: u8 = 1 << 2; // working-tree file edits → status, dirty
pub const INV_WORKTREES: u8 = 1 << 3; // .git/worktrees metadata
pub const INV_FULL: u8 = INV_REFS | INV_INDEX | INV_WORKDIR | INV_WORKTREES;

/// Typed refresh payloads pushed to the frontend on the `repo-event` channel.
#[derive(Serialize, Clone)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum RepoEvent {
    GraphChanged {
        gen: u64,
        head: Option<String>,
        commits: Vec<CommitNode>,
        unpushed: Vec<String>,
    },
    StatusChanged {
        gen: u64,
        status: WorkingStatus,
        dirty: bool,
    },
    WorktreesChanged {
        gen: u64,
        worktrees: Vec<Worktree>,
    },
    BranchesChanged {
        gen: u64,
        branches: Vec<String>,
    },
    RefreshError {
        gen: u64,
        op: String,
        message: String,
    },
}

/// Envelope so the frontend can discard events from a previously open repo.
#[derive(Serialize)]
struct Envelope<'a> {
    root: &'a str,
    #[serde(flatten)]
    event: &'a RepoEvent,
}

pub struct RepoService {
    tx: mpsc::UnboundedSender<u8>,
}

impl RepoService {
    /// Spawn the coordinator for `root`. Dropping the returned service closes
    /// the channel, which ends the coordinator task.
    pub fn start(app: tauri::AppHandle, root: String) -> Arc<Self> {
        let (tx, rx) = mpsc::unbounded_channel();
        tauri::async_runtime::spawn(coordinator(app, root, rx));
        Arc::new(Self { tx })
    }

    /// Queue an invalidation; the coordinator coalesces bursts into one cycle.
    pub fn poke(&self, bits: u8) {
        let _ = self.tx.send(bits);
    }
}

/// Last emitted values, so unchanged slices emit nothing (no spurious
/// re-renders on the frontend).
#[derive(Default)]
struct Cache {
    graph: Option<(Option<String>, Vec<CommitNode>, Vec<String>)>,
    status: Option<WorkingStatus>,
    worktrees: Option<Vec<Worktree>>,
    branches: Option<Vec<String>>,
}

async fn coordinator(app: tauri::AppHandle, root: String, mut rx: mpsc::UnboundedReceiver<u8>) {
    let gen = AtomicU64::new(0);
    let mut cache = Cache::default();

    while let Some(first) = rx.recv().await {
        let mut bits = first;
        // Coalesce: keep draining while events arrive in quick succession, but
        // never sit on a change for more than ~350ms total. Direct user
        // actions (a lone stage click) therefore land in ~80ms, while an
        // agent's save-storm collapses into one cycle.
        let started = Instant::now();
        loop {
            match tokio::time::timeout(Duration::from_millis(80), rx.recv()).await {
                Ok(Some(b)) => {
                    bits |= b;
                    if started.elapsed() > Duration::from_millis(350) {
                        break;
                    }
                }
                Ok(None) => return, // service dropped: stop the task
                Err(_) => break,    // quiet for 80ms: run the cycle
            }
        }
        let g = gen.fetch_add(1, Ordering::SeqCst) + 1;
        run_cycle(&app, &root, bits, g, &mut cache).await;
        // Invalidations that arrived mid-cycle are still queued and start the
        // next cycle immediately, so nothing is ever lost — only coalesced.
    }
}

async fn run_cycle(app: &tauri::AppHandle, root: &str, bits: u8, gen: u64, cache: &mut Cache) {
    let want_graph = bits & INV_REFS != 0;
    let want_status = bits & (INV_INDEX | INV_WORKDIR) != 0;
    let want_worktrees = bits & (INV_REFS | INV_INDEX | INV_WORKTREES) != 0;
    let want_branches = bits & INV_REFS != 0;

    // Kick off every dirty slice at once; each runs on the blocking pool.
    let graph_task = want_graph.then(|| {
        let p = root.to_string();
        tauri::async_runtime::spawn_blocking(move || {
            let commits = read::graph(&p, 400, None)?;
            let unpushed = read::unpushed_commits(&p).unwrap_or_default();
            let head = read::open(&p).ok().and_then(|r| r.head);
            anyhow::Ok((head, commits, unpushed))
        })
    });
    let status_task = want_status.then(|| {
        let p = root.to_string();
        tauri::async_runtime::spawn_blocking(move || read::working_status(&p))
    });
    let worktrees_task = want_worktrees.then(|| {
        let p = root.to_string();
        tauri::async_runtime::spawn_blocking(move || read::worktrees(&p))
    });
    let branches_task = want_branches.then(|| {
        let p = root.to_string();
        tauri::async_runtime::spawn_blocking(move || read::branches(&p))
    });

    if let Some(t) = graph_task {
        match flatten(t.await) {
            Ok(v) => {
                if cache.graph.as_ref() != Some(&v) {
                    emit(
                        app,
                        root,
                        &RepoEvent::GraphChanged {
                            gen,
                            head: v.0.clone(),
                            commits: v.1.clone(),
                            unpushed: v.2.clone(),
                        },
                    );
                    cache.graph = Some(v);
                }
            }
            Err(e) => emit_err(app, root, gen, "graph", e),
        }
    }
    if let Some(t) = status_task {
        match flatten(t.await) {
            Ok(s) => {
                if cache.status.as_ref() != Some(&s) {
                    let dirty =
                        !(s.staged.is_empty() && s.unstaged.is_empty() && s.untracked.is_empty());
                    emit(
                        app,
                        root,
                        &RepoEvent::StatusChanged {
                            gen,
                            status: s.clone(),
                            dirty,
                        },
                    );
                    cache.status = Some(s);
                }
            }
            Err(e) => emit_err(app, root, gen, "status", e),
        }
    }
    if let Some(t) = worktrees_task {
        match flatten(t.await) {
            Ok(w) => {
                if cache.worktrees.as_ref() != Some(&w) {
                    emit(
                        app,
                        root,
                        &RepoEvent::WorktreesChanged {
                            gen,
                            worktrees: w.clone(),
                        },
                    );
                    cache.worktrees = Some(w);
                }
            }
            Err(e) => emit_err(app, root, gen, "worktrees", e),
        }
    }
    if let Some(t) = branches_task {
        match flatten(t.await) {
            Ok(b) => {
                if cache.branches.as_ref() != Some(&b) {
                    emit(
                        app,
                        root,
                        &RepoEvent::BranchesChanged {
                            gen,
                            branches: b.clone(),
                        },
                    );
                    cache.branches = Some(b);
                }
            }
            Err(e) => emit_err(app, root, gen, "branches", e),
        }
    }
}

/// Collapse a JoinError or a slice error into one anyhow error.
fn flatten<T>(joined: Result<anyhow::Result<T>, tauri::Error>) -> anyhow::Result<T> {
    match joined {
        Ok(inner) => inner,
        Err(e) => Err(anyhow::anyhow!("task join failed: {e}")),
    }
}

fn emit(app: &tauri::AppHandle, root: &str, event: &RepoEvent) {
    let _ = app.emit("repo-event", &Envelope { root, event });
}

fn emit_err(app: &tauri::AppHandle, root: &str, gen: u64, op: &str, e: anyhow::Error) {
    emit(
        app,
        root,
        &RepoEvent::RefreshError {
            gen,
            op: op.to_string(),
            message: e.to_string(),
        },
    );
}

File: engine/src/service.ts
/**
 * Repo-state service: the single source of truth for "when does state refresh".
 *
 * One coordinator per watched repo turns classified invalidations (from the
 * file watcher, or from Grove's own write commands) into recomputed state
 * slices, compares them against the last emitted values, and pushes typed
 * events out. Every payload carries a generation number so a consumer can drop
 * anything stale. Nothing else re-fetches on its own; every refresh trigger
 * funnels through `poke`.
 *
 * Ported from `src-tauri/src/repo/service.rs`. The `AppHandle` the Rust version
 * took purely to call `emit` is now an `EmitFn` passed in by the caller, which
 * is what keeps this package free of any UI framework.
 */

import { isDeepStrictEqual } from 'node:util';

import * as read from './read.ts';
import {
  INV_INDEX,
  INV_REFS,
  INV_WORKDIR,
  INV_WORKTREES,
  type CommitNode,
  type RepoEvent,
  type RepoEventEnvelope,
  type WorkingStatus,
  type Worktree,
} from './types.ts';

/** How the coordinator hands finished events to its owner. */
export type EmitFn = (event: RepoEventEnvelope) => void;

/**
 * Coalescing window. A burst is drained while events keep arriving, but a
 * change is never sat on for more than ~350ms total. Direct user actions (a
 * lone stage click) therefore land in ~80ms, while an agent's save-storm
 * collapses into one cycle.
 */
const QUIET_MS = 80;
const MAX_COALESCE_MS = 350;

type RecvResult = { kind: 'value'; bits: number } | { kind: 'timeout' } | { kind: 'closed' };

/**
 * Unbounded single-consumer channel with a timed receive — the piece Rust got
 * from `tokio::sync::mpsc` plus `tokio::time::timeout`.
 */
class Channel {
  #queue: number[] = [];
  #waiter: ((r: RecvResult) => void) | null = null;
  #closed = false;

  send(bits: number): void {
    if (this.#closed) return;
    const w = this.#waiter;
    if (w) {
      this.#waiter = null;
      w({ kind: 'value', bits });
      return;
    }
    this.#queue.push(bits);
  }

  close(): void {
    this.#closed = true;
    const w = this.#waiter;
    if (w) {
      this.#waiter = null;
      w({ kind: 'closed' });
    }
  }

  recv(timeoutMs?: number): Promise<RecvResult> {
    const queued = this.#queue.shift();
    if (queued !== undefined) return Promise.resolve({ kind: 'value', bits: queued });
    if (this.#closed) return Promise.resolve({ kind: 'closed' });

    return new Promise<RecvResult>((resolve) => {
      let timer: NodeJS.Timeout | undefined;
      const settle = (r: RecvResult) => {
        if (timer) clearTimeout(timer);
        resolve(r);
      };
      this.#waiter = settle;

      if (timeoutMs !== undefined) {
        timer = setTimeout(() => {
          if (this.#waiter === settle) this.#waiter = null;
          settle({ kind: 'timeout' });
        }, timeoutMs);
        // A pending coalesce window should not by itself keep the process alive.
        timer.unref?.();
      }
    });
  }
}

/**
 * Last emitted values, so unchanged slices emit nothing (no spurious
 * re-renders downstream).
 */
interface Cache {
  graph?: { head: string | null; commits: CommitNode[]; unpushed: string[] };
  status?: WorkingStatus;
  worktrees?: Worktree[];
  branches?: string[];
}

type Settled<T> = { ok: true; value: T } | { ok: false; error: unknown };

function settle<T>(p: Promise<T>): Promise<Settled<T>> {
  return p.then(
    (value) => ({ ok: true as const, value }),
    (error: unknown) => ({ ok: false as const, error }),
  );
}

const errText = (e: unknown) => (e instanceof Error ? e.message : String(e));

export interface RepoService {
  /** Queue an invalidation; the coordinator coalesces bursts into one cycle. */
  poke(bits: number): void;
  /** Stop the coordinator. Safe to call more than once. */
  stop(): void;
}

/**
 * Start the coordinator for `root`. Calling `stop()` closes the channel, which
 * ends the coordinator loop.
 */
export function startRepoService(root: string, emit: EmitFn): RepoService {
  const ch = new Channel();
  void coordinator(root, ch, emit);
  return {
    poke: (bits) => ch.send(bits),
    stop: () => ch.close(),
  };
}

async function coordinator(root: string, ch: Channel, emit: EmitFn): Promise<void> {
  let gen = 0;
  const cache: Cache = {};

  for (;;) {
    const first = await ch.recv();
    if (first.kind === 'closed') return;
    if (first.kind === 'timeout') continue;

    let bits = first.bits;
    const started = Date.now();
    for (;;) {
      const next = await ch.recv(QUIET_MS);
      if (next.kind === 'closed') return; // service stopped: end the loop
      if (next.kind === 'timeout') break; // quiet for QUIET_MS: run the cycle
      bits |= next.bits;
      if (Date.now() - started > MAX_COALESCE_MS) break;
    }

    gen += 1;
    await runCycle(root, bits, gen, cache, emit);
    // Invalidations that arrived mid-cycle are still queued and start the next
    // cycle immediately, so nothing is ever lost — only coalesced.
  }
}

async function runCycle(
  root: string,
  bits: number,
  gen: number,
  cache: Cache,
  emit: EmitFn,
): Promise<void> {
  const wantGraph = (bits & INV_REFS) !== 0;
  const wantStatus = (bits & (INV_INDEX | INV_WORKDIR)) !== 0;
  const wantWorktrees = (bits & (INV_REFS | INV_INDEX | INV_WORKTREES)) !== 0;
  const wantBranches = (bits & INV_REFS) !== 0;

  const send = (event: RepoEvent) => emit({ ...event, root } as RepoEventEnvelope);
  const sendErr = (op: string, e: unknown) =>
    send({ kind: 'refresh_error', gen, op, message: errText(e) });

  // Kick off every dirty slice at once, then consume the results in order.
  const graphTask = wantGraph
    ? settle(
        (async () => {
          const commits = await read.graph(root, 400, null);
          const unpushed = await read.unpushedCommits(root).catch(() => []);
          const head = await read
            .open(root)
            .then((r) => r.head)
            .catch(() => null);
          return { head, commits, unpushed };
        })(),
      )
    : null;
  const statusTask = wantStatus ? settle(read.workingStatus(root)) : null;
  const worktreesTask = wantWorktrees ? settle(read.worktrees(root)) : null;
  const branchesTask = wantBranches ? settle(read.branches(root)) : null;

  if (graphTask) {
    const r = await graphTask;
    if (!r.ok) sendErr('graph', r.error);
    else if (!isDeepStrictEqual(cache.graph, r.value)) {
      cache.graph = r.value;
      send({
        kind: 'graph_changed',
        gen,
        head: r.value.head,
        commits: r.value.commits,
        unpushed: r.value.unpushed,
      });
    }
  }

  if (statusTask) {
    const r = await statusTask;
    if (!r.ok) sendErr('status', r.error);
    else if (!isDeepStrictEqual(cache.status, r.value)) {
      cache.status = r.value;
      const s = r.value;
      const dirty = !(
        s.staged.length === 0 &&
        s.unstaged.length === 0 &&
        s.untracked.length === 0
      );
      send({ kind: 'status_changed', gen, status: s, dirty });
    }
  }

  if (worktreesTask) {
    const r = await worktreesTask;
    if (!r.ok) sendErr('worktrees', r.error);
    else if (!isDeepStrictEqual(cache.worktrees, r.value)) {
      cache.worktrees = r.value;
      send({ kind: 'worktrees_changed', gen, worktrees: r.value });
    }
  }

  if (branchesTask) {
    const r = await branchesTask;
    if (!r.ok) sendErr('branches', r.error);
    else if (!isDeepStrictEqual(cache.branches, r.value)) {
      cache.branches = r.value;
      send({ kind: 'branches_changed', gen, branches: r.value });
    }
  }
}
